crates/task-impls/src/consensus.rs

```diff
@@ -894,86 +736,31 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
-                        accumulator,
-                        timeout_accumulator: either::Left(timeout_accumulator),
-                        cur_view: vote.get_view_number(),
+                    debug!("Starting vote handle for view {:?}", vote.get_view_number());
+                    let info = AccumulatorInfo {
+                        public_key: self.public_key.clone(),
+                        membership: self.quorum_membership.clone(),
+                        view: vote.get_view_number(),
                         event_stream: self.event_stream.clone(),
                         id: self.id,
+                        registry: self.registry.clone(),
                     };
                     let name = "Quorum Vote Collection";
-                    let filter = FilterEvent(Arc::new(|event| {
-                        matches!(
-                            event,
-                            HotShotEvent::QuorumVoteRecv(_) | HotShotEvent::TimeoutVoteRecv(_)
-                        )
-                    }));
-
-                    let builder =
-                        TaskBuilder::<VoteCollectionTypes<TYPES, I>>::new(name.to_string())
-                            .register_event_stream(self.event_stream.clone(), filter)
-                            .await
-                            .register_registry(&mut self.registry.clone())
-                            .await
-                            .register_state(state)
-                            .register_event_handler(handle_event);
-                    let id = builder.get_task_id().unwrap();
-                    let stream_id = builder.get_stream_id().unwrap();
-
-                    self.vote_collector = Some((vote.get_view_number(), id, stream_id));
-
-                    let _task = async_spawn(async move {
-                        VoteCollectionTypes::build(builder).launch().await;
-                    });
-                    debug!("Starting vote handle for view {:?}", vote.get_view_number());
+                    self.vote_collector = spawn_vote_accumulator::<
+                        TYPES,
+                        QuorumVote<TYPES>,
+                        QuorumCertificate<TYPES>,
+                    >(
+                        &info, vote.clone(), event, name.to_string()
+                    )
+                    .await;
                 } else if let Some((_, _, stream_id)) = self.vote_collector {
                     self.event_stream
-                        .direct_message(stream_id, HotShotEvent::QuorumVoteRecv(vote))
+                        .direct_message(stream_id, HotShotEvent::QuorumVoteRecv(vote.clone()))
                         .await;
                 }
             }
-            HotShotEvent::TimeoutVoteRecv(vote) => {
-                if self
-                    .timeout_membership
-                    .get_leader(vote.get_view_number() + 1)
-                    != self.public_key
```

Comment:
Is the reason that we no longer check the leadership that it's done in `spawn_vote_accumulator` in the vote task? I think this is indeed cleaner but may cause unnecessary operations since `spawn_vote_accumulator` is called at the end of this event handling after other checks. Would it be better if we keep the leadership check in individual tasks?

Reply:
I agree it's better to keep it in the individual tasks and remove it from the spawn function